src/refactor/ftp.py

```diff
@@ -0,0 +1,238 @@
+            self.operation = op
+        else:
+            raise ValueError("Invalid operation {}".format(op))
+
+    def setOutfile(self, file):
+        self.outfile = file
+
+    def setInfile(self, file):
+        self.infile = file
+
+    def _transaction(self, data):
+        try:
+            self.networkManager.send(self.satelliteAddr, self.destPort, data)
+            response = self.networkManager.receive(self.satelliteAddr, self.destPort, 10000)
+            return self.receiveParse.parseReturnValue(self.destPort, response)
+        except Exception as e:
+            print(e)
+            exit(1)
+
+    def _get_data_upload_packet(self, req_id, data, count):
+        subservice = self.services.get('FTP_COMMAND').get('subservice').get('FTP_UPLOAD_PACKET').get('subPort')
+        out = bytearray()
+        out.extend(subservice.to_bytes(1, byteorder='big'))
+        out.extend(req_id.to_bytes(4, byteorder='big'))
+        out.extend(count.to_bytes(4, byteorder='big'))
+        out.extend(len(data).to_bytes(4, byteorder='big'))
+        out.extend(bytearray(data))
+        return out
+
+    def _get_start_upload_packet(self, req_id, filesize):
+        subservice = self.services.get('FTP_COMMAND').get('subservice').get('FTP_START_UPLOAD').get('subPort')
+        out = bytearray()
+        out.extend(subservice.to_bytes(1, byteorder='big'))
+        out.extend(req_id.to_bytes(4, byteorder='big'))
+        out.extend(filesize.to_bytes(8, byteorder='big'))
+        out.extend(self.blocksize.to_bytes(4, byteorder='big'))
+        out.extend(bytes(self.outfile.encode("ascii")))
+        out.extend(int(0).to_bytes(1, byteorder='big'))
+        return out
+
+    def _get_file_size_packet(self):
+        subservice = self.services.get('FTP_COMMAND').get('subservice').get('GET_FILE_SIZE').get('subPort')
+        out = bytearray()
+        out.extend(subservice.to_bytes(1, byteorder='big'))
+        out.extend(bytes(self.infile.encode("ascii")))
+        out.extend(int(0).to_bytes(1, byteorder='big'))
+        return out
+
+    def _get_burst_download_packet(self, req_id, skip, count):     
+        subservice = self.services.get('FTP_COMMAND').get('subservice').get('REQUEST_BURST_DOWNLOAD').get('subPort')
+        out = bytearray()
+        out.extend(subservice.to_bytes(1, byteorder='big'))
+        out.extend(req_id.to_bytes(4, byteorder='big'))
+        out.extend(self.blocksize.to_bytes(4, byteorder='big'))
+        out.extend(int(skip).to_bytes(4, byteorder='big'))
+        out.extend(int(count).to_bytes(4, byteorder='big'))
+        out.extend(int(self.dest_addr).to_bytes(4, byteorder='big'))
+        out.extend(bytes(self.infile.encode("ascii")))
+        out.extend(int(0).to_bytes(1, byteorder='big'))
+        return out
+
+    def _receive_burst(self, req_id, skip, count):
+        if not os.path.exists(self.outfile):
+            mode = "wb"
+        else :
+            mode = "ab"
+        f = open(self.outfile, mode)
```

Comment:
The standard pythonic I/O loop is:

with open("foo") as f:
   #do file things
which magically calls close on exceptions
Would that work here?

Reply:
It would work, I just didn't understand it. I'll add it